cartography/intel/aws/permission_relationships.py

```diff
@@ -0,0 +1,200 @@
+                if evaluate_statment_clause_for_permission(statement, "resource", resource_arn):
+                    if not evaluate_statment_clause_for_permission(statement, "notresource", resource_arn):
+                        return True
+
+    return allowed
+
+
+def evaluate_policy_for_permission(statements, permissions, resource_arn):
+    # AWS Policy evaluation reference
+    # https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_evaluation-logic.html
+    # returns (allowed_by_policy, explicitly_denied_by_policy)
+    # return cases
+    # (True, False) - The policy allows the action
+    # (False, False) - The poliy does not allow the action, but also doesn't explicitly deny it
+    # (False, True) - The policy specifically denies the action. There is no need to evaluate other policies
+
+    allow_statements = [s for s in statements if s["effect"] == "Allow"]
+    deny_statements = [s for s in statements if s["effect"] == "Deny"]
+    for permission in permissions:
+        if evaluate_statements_for_permission(deny_statements, permission, resource_arn):
+            return False, True
+        else:
+            if evaluate_statements_for_permission(allow_statements, permission, resource_arn):
+                return True, False
+    return False, False
+
+
+def evaluate_policies_against_resource(policies, resource_arn, permissions):
+    granted = False
+    for policy_id, statements in policies.items():
+        allowed, explicit_deny = evaluate_policy_for_permission(statements, permissions, resource_arn)
+        # If the action is explicitly denied then no other policy can override it
+        if explicit_deny:
+            return False
+        if not granted and allowed:
+            granted = True
+
+    return granted
+
+
+def evaluate_relationships(principals, resource_arns, permission):
+    allowed_mappings = []
+    for resource_arn in resource_arns:
+        for principal_arn, policies in principals.items():
+            if evaluate_policies_against_resource(policies, resource_arn, permission):
+                allowed_mappings.append({"principal_arn": principal_arn, "resource_arn": resource_arn})
+    return allowed_mappings
+
+
+def parse_statement_node_group(node_group):
+    return [n._properties for n in node_group]
+
+
+def compile_regex(item):
+    if isinstance(item, str):
+        item = item.replace(".", "\\.").replace("*", ".*")
+        item = re.compile(item, flags=re.IGNORECASE)
+    return item
+
+
+def compile_statement(statements):
+    properties = ['action', 'resource', 'notresource', 'notaction']
+    for statement in statements:
+        for statement_property in properties:
+            if statement_property in statement:
+                statement[statement_property] = [compile_regex(item) for item in statement[statement_property]]
+    return statements
+
+
+def get_principals_for_account(neo4j_session, account_id):
```

Comment:
If possible we should avoid reading from the graph and then processing it through python because of the IO overhead from transferring stuff between Neo4j and python. Instead, we should try to keep all of this logic in Neo4j itself by merging on principal IDs.